### PR TITLE
Example: Function Calling in Typescript

### PR DESCRIPTION
**Pull request by technovangelist**
Two examples here. One to list the characters in the first few pages of War and Peace. The other parses emails for events and addresses.

### CODE REVIEW

File: examples/typescript-functioncalling/readme.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Function calling
+
+![function calling 2023-11-16 16_12_58](https://github.com/jmorganca/ollama/assets/633681/a0acc247-9746-45ab-b325-b65dfbbee4fb)
+
+One of the most exciting features added to Large Language Models recently is 'function calling'. It's a bit of a confusing name. It's understandable if you think that means the model can call functions, but that's not what it means. Function Calling simply means that the output of the model is formatted in JSON, using a preconfigured schema, and uses the expected types. Then your code can use the output of the model and call functions with it.
```
Comment (BruceMacD) [inline examples/typescript-functioncalling/readme.md:5]:
```suggestion
One of the most exciting features added to Large Language Models recently is 'function calling'. It's a bit of a confusing name. It's understandable if you think that means the model can call functions, but that's not what it means. Function calling simply means that the output of the model is formatted in JSON, using a preconfigured schema, and uses the expected types. Then your code can use the output of the model and call functions with it.
```

File: examples/typescript-functioncalling/readme.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+5. Review the `info.txt` file.
+6. Run `tsx extractemail.ts`.
+
+## Review the Code
+
+Both examples do roughly the same thing with different source material. They both use the same system prompt, which tells the model to expect some instructions and a schema. Then we inject the schema into the prompt and generate an answer.
+
+The first example, `extractwp.ts`, outputs the resulting JSON to the console, listing the characters introduced at the start of War and Peace. The second example, `extractemail.ts`, is a bit more complicated, extracting two different types of information: addresses and events. It outputs the results to a JSON blob, then the addresses are handed off to one function called `reportAddresses` and the events are handed off to another function called `reportEvents`.
+
+Notice that both examples are using the model from Intel called `neural-chat`. This is not a model tuned for function calling, yet it performs very well at this task.
```
Comment (BruceMacD) [inline examples/typescript-functioncalling/readme.md:24]: What was the reasoning for choosing to use `neural-chat`? I'm curious as a reader.

File: examples/typescript-functioncalling/extractwp.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        "description": "Name of the person"
+      },
+      "title": {
+        "type": "string",
+        "description": "Title of the person"
+      }
+    }],
+  }
+
+  // Depending on the model chosen, you may have a restricted context.
```
Comment (BruceMacD) [inline examples/typescript-functioncalling/extractwp.ts:24]:
```suggestion
  // Depending on the model chosen, you may be limited by the size of the context window, so limit the context to 2000 words.
```

**Review by BruceMacD — COMMENTED** [review]
Really cool example, minor feedback

File: examples/typescript-functioncalling/readme.md (hunk 2 above)
Comment (technovangelist) [reply examples/typescript-functioncalling/readme.md:24]: it was new and shiny

**Comment by technovangelist** [comment]
anything else @BruceMacD

**Review by BruceMacD — APPROVED** [review]
looks good to me

**Pull request merged**